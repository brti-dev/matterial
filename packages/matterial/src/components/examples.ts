/**
 * Examples JSX components
 * Not exported to the main package
 * Used for docs and tests
 */

export * from './Avatar/avatar.examples'
export * from './Button/button.examples'
export * from './Checkbox/checkbox.examples'
export * from './CheckButton/check-button.examples'
export * from './Dialog/dialog.examples'
