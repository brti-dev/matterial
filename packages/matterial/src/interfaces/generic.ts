/**
 * Remove properties `K` from `T`.
 * Distributive for union types.
 */
export type DistributiveOmit<T, K extends keyof any> = T extends any
  ? Omit<T, K>
  : never
