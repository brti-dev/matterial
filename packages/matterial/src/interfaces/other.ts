export type DateTimeType = Date | string | number

/**
 * Component with aria label
 * @usage type Props = Props_base & (AriaLabelledBy | AriaLabel)
 */
export type AriaLabel = {
  label: string
  labelledBy?: never
}
export type AriaLabelledBy = {
  labelledBy: string
  label?: never
}

export type InsertContent = {
  /**
   * Content to put on the RIGHT side of children/main content
   */
  append?: React.ReactNode

  /**
   * Content to put on the LEFT side of children/main content
   */
  prepend?: React.ReactNode
}
