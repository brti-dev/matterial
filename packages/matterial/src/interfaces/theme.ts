export type Severity = 'error' | 'warning' | 'info' | 'success'

export type Color =
  | Severity
  | 'default'
  | 'primary'
  | 'secondary'
  | 'red'
  | 'green'
  | 'dark'
  | 'light'

export type Variant = 'default' | 'contained' | 'outlined'
