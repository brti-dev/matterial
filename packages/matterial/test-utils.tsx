import { FC, ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

// Example of wrapping a test
function ThemeProvider({ theme, children }: any) {
  return <div className={`theme--${theme}`}>{children}</div>
}

const AllTheProviders: FC = ({ children }: any) => {
  return <ThemeProvider theme="light">{children}</ThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'
export { customRender as render }
