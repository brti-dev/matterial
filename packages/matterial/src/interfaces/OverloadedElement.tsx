import { DistributiveOmit } from './generic'

/**
 * Elements that accept no children
 */
type VoidElement =
  | 'area'
  | 'base'
  | 'br'
  | 'col'
  | 'hr'
  | 'img'
  | 'input'
  | 'link'
  | 'meta'
  | 'param'
  | 'command'
  | 'keygen'
  | 'source'

type OmitChildrenFromVoid<C extends React.ElementType> = C extends VoidElement
  ? Omit<React.ComponentPropsWithRef<C>, 'children'>
  : React.ComponentPropsWithRef<C>

export interface OverloadedElement<P> {
  <C extends React.ElementType>(
    props: { as: C } & P & DistributiveOmit<OmitChildrenFromVoid<C>, keyof P>
  ): JSX.Element
  (
    props: P & DistributiveOmit<React.ComponentPropsWithRef<'div'>, keyof P>
  ): JSX.Element
}

export type OverloadedElementProps = {
  as?: string | React.ElementType
}
