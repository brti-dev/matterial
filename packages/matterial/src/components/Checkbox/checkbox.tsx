import * as React from 'react'

import { ChangeEvent, CheckboxChangeEvent } from '../Form'
import { ColoredElement } from '../../interfaces/theme'
import { Icon } from '../Icon'
import classnames from '../../lib/classnames'
import useColor from '../../lib/use-color'
import classes from './checkbox.module.scss'
import { RequiredChildren } from '../../interfaces/children'

type Size = number | 'small' | 'medium' | 'large'

type Props = RequiredChildren &
  ColoredElement & {
    /**
     * Indicates if it's initially checked or not; **Controlled component**:
     * state is only set by the `checked` prop, not by the DOM
     */
    checked?: boolean
    /**
     * Indicates the checkbox is not changeable
     */
    disabled?: boolean
    /**
     * Limbo state
     */
    indeterminate?: boolean
    /**
     * Input name
     */
    name: string
    /**
     * Callback executed when the input value changes
     */
    onChange?: CheckboxChangeEvent
    /**
     * Size of checkbox and label; Number in pixels
     */
    size?: Size
  }

type NativeProps = Omit<React.ComponentPropsWithRef<'input'>, keyof Props>
type CheckboxProps = NativeProps & Props

export const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (props: CheckboxProps, ref: any) => {
    const {
      checked: controlledChecked,
      children,
      className,
      color: naturalColor = 'default',
      disabled = false,
      name,
      onChange = () => null,
      size = 'medium',
      style: naturalStyle,
    } = props
    const color = useColor(naturalColor)
    const sizePx = typeof size === 'number' ? `${size}px` : null
    const sizeClass = !sizePx ? `size--${size}` : null
    const style = {
      ...naturalStyle,
      ...color.style,
      ...(sizePx && { '--size': sizePx }),
    }
    const classNames = classnames(
      classes.checkbox,
      className,
      sizeClass,
      `variant--default`,
      color.className,
      disabled ? classes.disabled : undefined
    )

    // Provide local state if the parent is not controlling this component
    const isControlled = controlledChecked !== undefined
    const [localChecked, setLocalChecked] = React.useState<boolean>(
      controlledChecked || false
    )

    const checked = isControlled ? controlledChecked : localChecked

    const handleChange = () => {
      if (isControlled) {
        onChange(name, !controlledChecked)
      } else {
        setLocalChecked(!localChecked)
        onChange(name, localChecked)
      }
    }

    return (
      <label className={classNames} style={style}>
        <input
          type="checkbox"
          name={name}
          className="visually-hidden"
          ref={ref}
          onChange={handleChange}
          checked={checked}
          disabled={disabled}
        />
        <CheckboxIcon checked={checked} indeterminate={props.indeterminate} />
        <span className={classes.label}>{children}</span>
      </label>
    )
  }
)

function CheckboxIcon({
  checked,
  indeterminate,
}: Pick<CheckboxProps, 'checked' | 'indeterminate'>): React.ReactElement {
  let icon: string
  if (checked) {
    icon = 'CheckboxChecked'
  } else if (indeterminate) {
    icon = 'CheckboxMinus'
  } else {
    icon = 'Checkbox'
  }

  return <Icon icon={icon} aria-hidden />
}
