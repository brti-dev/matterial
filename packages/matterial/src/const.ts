import { Color, Accent } from './interfaces/theme'

export const COLORS: Color[] = [
  'primary',
  'secondary',
  'error',
  'warning',
  'info',
  'success',
  'dark',
  'light',
]

export const ACCENTS: Accent[] = [
  'accent-1',
  'accent-2',
  'accent-3',
  'accent-4',
  'accent-5',
  'accent-6',
  'accent-7',
  'accent-8',
]
