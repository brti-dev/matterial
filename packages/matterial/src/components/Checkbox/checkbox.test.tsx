import * as React from 'react'
import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event'

import { render, screen } from '../../test-utils'
import { Color } from '../../interfaces/theme'
import { COLORS } from '../../const'
import { Checkbox } from '.'
import { CheckboxExample } from './checkbox.examples'

describe('Checkbox component', () => {
  test('should render correctly', () => {
    const tree = renderer
      .create(
        <Checkbox name="checkbox" color="primary" className="foo" checked>
          Checkbox
        </Checkbox>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should work with different colors', () => {
    const tree = renderer
      .create(
        <>
          {[...COLORS, 'hotpink'].map(color => (
            <Checkbox name="checkbox" color={color} key={color}>
              {color}
            </Checkbox>
          ))}
        </>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should work with different sizes', () => {
    const tree = renderer
      .create(
        <>
          {['small', 'large', 50].map((size: any) => (
            <Checkbox name="checkbox" size={size} key={size}>
              {size}
            </Checkbox>
          ))}
        </>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render checked initially and toggle', () => {
    render(<CheckboxExample />)

    const cbox = screen.getByLabelText("I'm a daddy")
    expect(cbox).toBeChecked()

    // @ts-ignore
    userEvent.click(cbox)
    expect(cbox).not.toBeChecked()
  })

  test('should trigger a callback onChange', () => {
    const CheckboxComponent = () => {
      const [change, setChange] = React.useState(false)

      return (
        <Checkbox name="checkbox" onChange={() => setChange(true)}>
          {change ? 'is checked' : 'not checked'}
        </Checkbox>
      )
    }

    render(<CheckboxComponent />)
    const cbox = screen.getByRole('checkbox')
    userEvent.click(cbox)
    expect(screen.getByText('is checked')).toBeInTheDocument()
  })

  test('should be disabled', () => {
    render(
      <Checkbox name="checkbox" disabled>
        disabled
      </Checkbox>
    )

    expect(screen.getByRole('checkbox')).toBeDisabled()
  })
})
