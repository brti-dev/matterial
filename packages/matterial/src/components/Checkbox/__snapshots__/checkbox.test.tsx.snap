// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component should render correctly 1`] = `
<label
  className="checkbox foo size--medium variant--default color--primary"
  style={
    Object {
      "--color": "var(--color-primary)",
    }
  }
>
  <input
    checked={true}
    className="visually-hidden"
    onChange={[Function]}
    type="checkbox"
  />
  <svg
    aria-hidden={true}
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 24 24"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
    />
    <path
      d="M10.996 12.556 9.7 11.285l-1.4 1.43 2.704 2.647 4.699-4.651-1.406-1.422z"
    />
  </svg>
  <span
    className="label"
  >
    Checkbox
  </span>
</label>
`;

exports[`Checkbox component should work with different colors 1`] = `
Array [
  <label
    className="checkbox size--medium variant--default color--primary"
    style={
      Object {
        "--color": "var(--color-primary)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      primary
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--secondary"
    style={
      Object {
        "--color": "var(--color-secondary)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      secondary
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--dark"
    style={
      Object {
        "--color": "var(--color-dark)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      dark
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--light"
    style={
      Object {
        "--color": "var(--color-light)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      light
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--error"
    style={
      Object {
        "--color": "var(--color-error)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      error
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--warning"
    style={
      Object {
        "--color": "var(--color-warning)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      warning
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--info"
    style={
      Object {
        "--color": "var(--color-info)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      info
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--success"
    style={
      Object {
        "--color": "var(--color-success)",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      success
    </span>
  </label>,
  <label
    className="checkbox size--medium variant--default color--hotpink"
    style={
      Object {
        "--color": "hotpink",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      hotpink
    </span>
  </label>,
]
`;

exports[`Checkbox component should work with different sizes 1`] = `
Array [
  <label
    className="checkbox size--small variant--default color--default"
    style={
      Object {
        "--color": null,
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      small
    </span>
  </label>,
  <label
    className="checkbox size--large variant--default color--default"
    style={
      Object {
        "--color": null,
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      large
    </span>
  </label>,
  <label
    className="checkbox variant--default color--default"
    style={
      Object {
        "--color": null,
        "--size": "50px",
      }
    }
  >
    <input
      className="visually-hidden"
      onChange={[Function]}
      type="checkbox"
    />
    <svg
      aria-hidden={true}
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 5c-1.103 0-2 .897-2 2v10c0 1.103.897 2 2 2h10c1.103 0 2-.897 2-2V7c0-1.103-.897-2-2-2H7zm0 12V7h10l.002 10H7z"
      />
    </svg>
    <span
      className="label"
    >
      50
    </span>
  </label>,
]
`;
